File: src/main/java/net/dontdrinkandroot/gitki/model/FilePath.java
package net.dontdrinkandroot.gitki.model;

import java.nio.file.Path;
import java.util.Arrays;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class FilePath extends AbstractPath
{
    FilePath(String name, DirectoryPath directoryPath)
    {
        this.setName(name);
        this.setParent(directoryPath);
    }

    public FilePath(String name)
    {
        this.setName(name);
        this.setParent(new DirectoryPath());
    }

    public static FilePath parse(String pathString)
    {
        if (pathString.endsWith(SEPARATOR)) {
            throw new IllegalArgumentException("pathString must not end with " + SEPARATOR);
        }

        String[] parts = pathString.split(SEPARATOR);

        FilePath filePath = new FilePath(parts[parts.length - 1]);

        if (parts.length > 1) {
            filePath.setParent(DirectoryPath.from(Arrays.copyOfRange(parts, 0, parts.length - 1)));
        }

        return filePath;
    }

    @Override
    public String toString()
    {
        return this.getParent().toString() + this.getName();
    }

    @Override
    public Path toPath()
    {
        Path parentPath = this.getParent().toPath();
        return parentPath.resolve(this.getName());
    }

    @Override
    public boolean isRoot()
    {
        return false;
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/model/FilePath.kt
package net.dontdrinkandroot.gitki.model

import net.dontdrinkandroot.gitki.model.DirectoryPath.Companion.from
import org.apache.commons.io.FilenameUtils
import java.nio.file.Path as NioPath

class FilePath : AbstractPath {

    internal constructor(name: String, directoryPath: DirectoryPath) {
        this.name = name
        parent = directoryPath
    }

    constructor(name: String) {
        this.name = name
        parent = DirectoryPath()
    }

    override fun toString(): String = parent.toString() + name

    override val nioPath: NioPath
        get() {
            val parentPath = parent!!.nioPath
            return parentPath.resolve(name)
        }

    override val root: Boolean
        get() = false

    override val directoryPath = false

    val extension: String
        get() = FilenameUtils.getExtension(name)

    val nameWithoutExtension: String
        get() = FilenameUtils.removeExtension(name)

    companion object {

        @JvmStatic
        fun parse(pathString: String): FilePath {
            require(!pathString.endsWith(SEPARATOR)) { "pathString must not end with $SEPARATOR" }
            val parts: List<String> = pathString.split(SEPARATOR)
            val filePath = FilePath(parts.last())
            if (parts.size > 1) {
                filePath.parent = from(parts.subList(0, parts.size - 1))
            }
            return filePath
        }

        @JvmStatic
        fun from(path: NioPath): FilePath {
            var directoryPath = DirectoryPath()
            val parent = path.parent
            if (null != parent) {
                directoryPath = DirectoryPath.from(path.parent)
            }
            return directoryPath.appendFileName(path.fileName.toString())
        }
    }
}